File: javascript/15.3-sum.js
/*
 * @lc app=leetcode id=15 lang=javascript
 *
 * [15] 3Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  let tripletSum = []
  if (nums.length < 3) return tripletSum
  // sort and shift 3 pointers
  // O(nlog(n) + n^2)
  nums.sort((a, b) => a - b)
  for (let i = 0; i < nums.length - 2; i++) {
    // can't ever reach a sum of 0
    if (nums[i] > 0) return tripletSum
    // avoid checking repeated digits
    if (i > 0 && nums[i] === nums[i-1]) continue
    let j = i + 1
    let k = nums.length - 1
    while (j < k) {
      let sum = nums[i] + nums[j] + nums[k]
      if (sum === 0) {
        tripletSum.push([nums[i], nums[j], nums[k]])
        k--
        j++
      } else if (sum > 0) {
        k--
        continue
      } else {
        j++
        continue
      }
      // avoid checking repeated digits
      while (nums[k] === nums[k+1]) {
        k--
      }
      while (nums[j] === nums[j-1]) {
        j++
      }
    }
  }
  return tripletSum
};
// @lc code=end


console.log(threeSum([-1,0,1,2,-1,-4]))
console.log(threeSum([-1,0,1,2,-1,-4,-2,-3,3,0,4]))
console.log(threeSum([-2,0,1,1,2]))

File: cpp/15.3-sum.cpp
/*
 * @lc app=leetcode id=15 lang=cpp
 *
 * [15] 3Sum
 */

// @lc code=start
class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    {
        // brute force approach
        // iterate over every 3 pair combination
        // O(n^3) time O(k) space
        // sorted approach
        // sort, then iterate once, shift indexes
        // high and low based on curr sum
        // O(nlog(n) + n^2) time O(k) space

        vector<vector<int>> resVect;

        if (nums.size() < 3)
            return resVect;

        // sort input
        sort(nums.begin(), nums.end());

        int i = 0;
        for (; i < nums.size(); i++)
        {
            int pLow = i + 1;
            int pHigh = nums.size() - 1;
            // eject early out of repeating digits
            if (i > 0 && nums[pLow] == nums[pHigh])
            {
                break;
            }
            while (pHigh > pLow)
            {
                int currSum = nums[i] + nums[pLow] + nums[pHigh];
                if (currSum == 0)
                {
                    vector<int> triplet{nums[i], nums[pLow], nums[pHigh]};
                    resVect.push_back(triplet);
                    // keep going!
                    pHigh--;
                }
                else if (currSum > 0)
                {
                    pHigh--;
                }
                else
                {
                    pLow++;
                }
            }
        }

        // c++ uniqueness in multidimensional vector magic
        sort(resVect.begin(), resVect.end());
        resVect.erase(unique(resVect.begin(), resVect.end()), resVect.end());

        return resVect;
    }
};
// @lc code=end
